### DISCUSSION TITLE
Font and Font Size Control

### DISCUSSION THREAD

**Discussion by rmills** [RudolphRiedel/FT800-FT813#42] [Q&A]
Trying to figure out how the library handles the fonts. _EVE_cmd_text_burst()_ takes _"int16_t font"_ but there is not reference to what that variable refers too? The examples only use 26 to 29, others seem to just draw boxes. I found some functions for loading fonts but nothing about controlling the size. 

Are the loaded font/fonts a fixed size or is there a way to control them in the display list? 

Is it possible to mix a very large font size with a small size?

**Comment by RudolphRiedel** [comment]
While you are correct that there is very little documentation included here, the reason is that most of the functions are implemented exactly as described in the FT81x / BT81x programming guides and they are not doing anything functional, they just provide a way to send a stream of bytes in the required format to the chip.

And Bridgetek provides the documentation:
https://brtchip.com/ic-module/document/programming-guides/
https://brtchip.com/ic-module/wp-content/uploads/sites/3/2021/10/FT81X_Series_Programmer_Guide.pdf
https://brtchip.com/ic-module/wp-content/uploads/sites/3/2021/10/BRT_AN_033_BT81X-Series-Programming-Guide.pdf

That said, what the number for the font actually refers to is a bitmap-handle and there 32 of them so the range is 0-31.
I really have no idea why the parameter is an int16_t, but this is what the command requires. 
By default the bitmap-handles 16 to 31 point to the ROM fonts and 26 to 31 are different sized versions of the same font.
All the fonts, ROM or custom, are a collection of bitmaps so there is no practical way to scale their size.

And the only way to mix different font sizes is to use separate commands.
There is no degree symbol in the ROM fonts for example, as in "21°C".
And the number of characters for FT8xx is limited to 127.
Instead of using a custom font and replacing some symbol I simply have a FT81x print "21 C" first and then "o" in a smaller font.
The tricky bit is just to figure out where to place it.

With BT81x though UTF-8 custom fonts are possible.

The CMD_SETFONT2 is used to register a custom font with a bitmap-handle.
And for this command the parameter to hold 0 to 31 has a type of uint32_t.
Most likely to avoid the need to shuffle bytes around to form 32 bit words for the other parameters.

**Comment by rmills** [comment]
Oh sorry, I had NewHavens docs and they were just different. Their library is implemented a lot differently. Looking at the BRT docs I understand why you have it the way you do. Looks like I need to create a new font, I'm on an FT812, I will have look and see if I can figure out that process. If I can figure out how to get a large font, I think I have the rest sorted out.

**Reply by RudolphRiedel** [reply]
How large are we talking about?
The FT81x have three more fonts numbered 32, 33 and 34, these only need to be assigned to a bitmap-handle, CMD_ROMFONT does this. Beyond these you need a custom font.

**Comment by Greg-R** [comment]
I am very confused with regards to the BT817 "ROM fonts".  Page 102 of the progamming guide says fonts 16 to 34 (except 17 and 19) contain only 95 printable ASCII characters, with 17 and 19 containing 127 printable ASCII characters.
I thought the BT817 has UTF-8?  Or is that only in the case of loading a custom font?
I am attempting to use command text to write to the display the non-ASCII characters ↑ and ↓ (up arrow and down arrow).
So my question is, do I need to load a custom font to use these symbols?

**Reply by RudolphRiedel** [reply]
Yes, the BT817has support for UTF-8 fonts and none of the built-in fonts even has Latin-1 Supplement or U+0080 to U+00FF, unfortunately. I would have gladly given up some of the other fonts for that.

The builtin font is Roboto-regular, but I have no idea where the arrows are located and if it even does have these.

Another set of free to use assets are the Google material icons: https://github.com/google/material-design-icons

**Reply by RudolphRiedel** [reply]
I just checked and Roboto does not have the arrows with are in U-2190 to U-21FF.
One of the fonts I like to use for a different reason has these: https://fonts.google.com/specimen/Inconsolata
And what I like about this one is that the zero has a slash thru it.
Another one with the arrows would be https://fonts.google.com/noto/specimen/Noto+Sans+Symbols

**Comment by Greg-R** [comment]
How much memory will be required in RAM_G to hold a custom font?

**Reply by RudolphRiedel** [reply]
That is not exactly easy to answer. This depends on the font itself, how big the resulting glyphs are for a given font size, what compression you use (I generally recommend ASTC 8x8), how many pages you select and how the different glyphs affect the glyph size.
I just converted Inconsolate to a size of 36 using Basic Latin, Latin-1 Supplement and Arrows, the size is 88 kiB for the .glyph file and the 204 glyphs are 32x40 in size.
Adding "Latin Extended A" with 124 glyphs changes the pixel size to 32x48 and the result is 153kiB for 328 glyphs.
Adding "Latin Extended B" with 60 glyphs changes the pixel size to 32x48 and the result is 177kiB for 264 glyphs.
Changing the size to 40 does not change anything, the size for 204 glyphs is still 88kiB.
Changing the size to 42 changes the pixel size to 32x48 and the result is 106kiB for 204 glyphs.

So you need to try what works for you and where you land with this.

**Reply by Greg-R** [reply]
OK, the memory requirements are not excessively high.  I will probably be able to fit at least one font in RAM_G.
